e3sm_diags/derivations/utils.py

```diff
@@ -0,0 +1,309 @@
+    if prs_low in low_bnds and prs_high in high_bnds:
+        return "middle cloud fraction"
+    elif prs_low in low_bnds:
+        return "high cloud fraction"
+    elif prs_high in high_bnds:
+        return "low cloud fraction"
+    else:
+        return "total cloud fraction"
+
+
+def cosp_bin_sum(
+    cld: "FileVariable",
+    prs_low0: Optional[float],
+    prs_high0: Optional[float],
+    tau_low0: Optional[float],
+    tau_high0: Optional[float],
+):
+    # FIXME: Refactor this function to operate on xr.Dataset/xr.DataArray.
+    """sum of cosp bins to calculate cloud fraction in specified cloud top pressure / height and
+    cloud thickness bins, input variable has dimension (cosp_prs,cosp_tau,lat,lon)/(cosp_ht,cosp_tau,lat,lon)
+    """
+    prs: FileAxis = cld.getAxis(0)
+    tau: FileAxis = cld.getAxis(1)
+
+    prs_low: float = adjust_prs_val_units(prs, prs[0], prs_low0)
+    prs_high: float = adjust_prs_val_units(prs, prs[-1], prs_high0)
+
+    if prs_low0 is None and prs_high0 is None:
+        prs_lim = "total cloud fraction"
+
+    tau_high, tau_low, tau_lim = determine_tau(tau, tau_low0, tau_high0)
+
+    if cld.id == "FISCCP1_COSP":  # ISCCP model
+        cld_bin = cld(cosp_prs=(prs_low, prs_high), cosp_tau=(tau_low, tau_high))
+        simulator = "ISCCP"
+    if cld.id == "CLISCCP":  # ISCCP obs
+        cld_bin = cld(isccp_prs=(prs_low, prs_high), isccp_tau=(tau_low, tau_high))
+
+    if cld.id == "CLMODIS":  # MODIS
+        prs_lim = determine_cloud_level(prs_low, prs_high, (440, 44000), (680, 68000))
+        simulator = "MODIS"
+
+        if prs.id == "cosp_prs":  # Model
+            cld_bin = cld(
+                cosp_prs=(prs_low, prs_high), cosp_tau_modis=(tau_low, tau_high)
+            )
+        elif prs.id == "modis_prs":  # Obs
+            cld_bin = cld(modis_prs=(prs_low, prs_high), modis_tau=(tau_low, tau_high))
+
+    if cld.id == "CLD_MISR":  # MISR model
+        if max(prs) > 1000:  # COSP v2 cosp_htmisr in units m instead of km as in v1
+            cld = cld[
+                1:, :, :, :
+            ]  # COSP v2 cosp_htmisr[0] equals to 0 instead of -99 as in v1, therefore cld needs to be masked manually
+        cld_bin = cld(cosp_htmisr=(prs_low, prs_high), cosp_tau=(tau_low, tau_high))
+        prs_lim = determine_cloud_level(prs_low, prs_high, (7, 7000), (3, 3000))
+        simulator = "MISR"
+    if cld.id == "CLMISR":  # MISR obs
+        cld_bin = cld(misr_cth=(prs_low, prs_high), misr_tau=(tau_low, tau_high))
+
+    cld_bin_sum = MV2.sum(MV2.sum(cld_bin, axis=1), axis=0)
+
+    try:
+        cld_bin_sum.long_name = simulator + ": " + prs_lim + " with " + tau_lim
+        # cld_bin_sum.long_name = "{}: {} with {}".format(simulator, prs_lim, tau_lim)
+    except BaseException:
+        pass
+    return cld_bin_sum
+
+
+def determine_tau(
+    tau: "FileAxis", tau_low0: Optional[float], tau_high0: Optional[float]
+):
+    # FIXME: Refactor this function to operate on xr.Dataset/xr.DataArray.
```

Comment:
Is looks like this still has to be done?

Reply:
Not yet, it needs to be refactored in https://github.com/E3SM-Project/e3sm_diags/issues/716.